### PR TITLE
REGD-121-chipdata

### PR DESCRIPTION
**Pull request by mingjiecn**
What did I do in this ticket:
1. add some example input for query form
2. adjust min distance between labels for snp diagram since I use bigger font
3. add search page header for search page. Click the button will switch views. Right now there are three view links in the header, will add more when create more views.
4. add chip data view. Add chip data table and ship data bar chart into chip data view.
5. add a blank accessibility data view component. Just for testing view switch.
6. create score data view area for displaying data item two in a row.

### CODE REVIEW

File: components/accessibility-view.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+  const router = useRouter();
+
+  useEffect(() => {
+    const isChip = router.asPath.endsWith(`#!accessibility`);
+    setShowChipData(isChip);
+  }, [router]);
+  return showChipData;
+}
+/**
+ * This is the view for diasplay accessibility data for a variant.
```
Comment (forresttanaka) [inline components/accessibility-view.js]: `display`

File: components/accessibility-view.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+export function useAccessibilityData() {
+  const [showChipData, setShowChipData] = useState(false);
+  const router = useRouter();
+
+  useEffect(() => {
+    const isChip = router.asPath.endsWith(`#!accessibility`);
+    setShowChipData(isChip);
+  }, [router]);
+  return showChipData;
+}
```
Comment (forresttanaka) [inline components/accessibility-view.js]: Impressed by you impelementing a custom hook — something a lot of devs don’t even think of using. Since you don’t use this custom hook anywhere else, make sure you can defend your use of it instead of just putting this code directly in `AccessibilityDataView`. Custom hooks are great for sharing React state and functionality among different components, but their utility gets reduced if you only use it in one component.

File: components/accessibility-view.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+          <DataPanel>
+            <DataAreaTitle>
+              No accessibility data available to display, please choose a
+              different SNP.
+            </DataAreaTitle>
+          </DataPanel>
+        </React.Fragment>
+      )}
+    </React.Fragment>
+  ) : null;
```
Comment (forresttanaka) [inline components/accessibility-view.js]: You can shorten this to:
```
return showData && (<React.Fragment>…</ReactFragment>)
```
…just to save a little typing (functionally nearly identical), and I think it makes the code look cleaner. And in these cases you can use `<>…</>` instead of `<React.Fragment></React.Fragment>`, but that’s personal preference — functionally identical.

File: components/chip-data-bar-chart.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+  const counts = chipData.reduce((groupCountBytarget, dataset) => {
+    const target = dataset.targets;
+    if (target in groupCountBytarget) {
+      groupCountBytarget[target] += 1;
+    } else {
+      groupCountBytarget[target] = 1;
+    }
+    return groupCountBytarget;
+  }, {});
+  console.log(counts);
```
Comment (forresttanaka) [inline components/chip-data-bar-chart.js]: You probably want to remove this? Would also be nice to see a description of what `counts` basically looks like at the end.

File: components/chip-data-bar-chart.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+  const labels = Object.keys(counts).sort((a, b) => {
+    return counts[b] - counts[a];
+  });
```
Comment (forresttanaka) [inline components/chip-data-bar-chart.js:49]: This is a simple case for `_.sortBy`.

File: components/chip-data-bar-chart.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+  const data = {
+    labels,
+    datasets: [
+      {
+        label: "Number of ChIP-seq datasets",
+        data: values,
+        backgroundColor: "#276A8E",
+      },
+    ],
+  };
```
Comment (forresttanaka) [inline components/chip-data-bar-chart.js:62]: Most of the section up through here looks like kind of a blur of code. That might be kind of OK if I was asked to maintain this as I could run it and console.log things to see how they work, but the lack of description makes code review take time and be not so useful to you.

File: components/chip-data-table.js [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+const chipDataColumns = [
+  {
+    id: "method",
+    title: "Method",
+  },
+  {
+    id: "peak",
+    title: "Peak",
+    display: ({ source }) => `${source.chrom}:${source.start}-${source.end}`,
+    sorter: (source) => Number(source.start),
```
Comment (forresttanaka) [inline components/chip-data-table.js:16]: Nice. We don’t often get a chance to use `sorter`.

File: components/chip-data-table.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+      source.biosample_ontology
+        ? source.biosample_ontology.organ_slims.join(", ")
+        : "",
+  },
+  {
+    id: "dataset",
+    title: "Dataset",
+    display: ({ source }) => (
+      <a href={source.dataset}>{source.dataset.split("/")[4]}</a>
+    ),
```
Comment (forresttanaka) [inline components/chip-data-table.js:45]: Just confirming, `source.dataset` is an external link? I don’t know what it is. If it’s an internal link, then this should use the `<Link>` component.

File: components/chip-data-view.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
+import dynamic from "next/dynamic";
+import { useRouter } from "next/router";
+import React, { useEffect, useState } from "react";
+import PropTypes from "prop-types";
+import ChipDataTable from "./chip-data-table";
+import { DataAreaTitle, DataPanel } from "./data-area";
+
+const ChipDataBar = dynamic(() => import("./chip-data-bar-chart"), {
+  ssr: false,
+});
```
Comment (forresttanaka) [inline components/chip-data-view.js:12]: Would you be able to describe how this works? If I was asked to fix issues with this, I wouldn’t even know why we need this at all, and I’d feel compelled to remove it in favor of a regular import, like I ended up doing successfully as a test with igvf-ui edit-func.

And I don’t mean describe it here. But maybe I should research this myself. I never found a reason to use it, so I’m not at all sure about it.

File: components/chip-data-view.js [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+export function useChipData() {
+  const [showChipData, setShowChipData] = useState(false);
+  const router = useRouter();
+
+  useEffect(() => {
+    const isChip = router.asPath.endsWith(`#!chip`);
+    setShowChipData(isChip);
+  }, [router]);
+  return showChipData;
+}
```
Comment (forresttanaka) [inline components/chip-data-view.js]: Like the other, make sure you can justify the custom hook instead of just using this code directly in `ChipDataView`, as this is the only place you use it.

File: components/chip-data-view.js [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+/**
+ * This is the view for diasplay ChIP-seq data for a variant.
+ */
+export function ChipDataView({ chipData }) {
+  const showChipData = useChipData();
+  return showChipData ? (
```
Comment (forresttanaka) [inline components/chip-data-view.js]: Another good place for the `return showChipData && <>…</>` mechanism. You don’t _have_ to, but it’s certainly common and clean.

File: components/score-view-data-area.js [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+    <div className="@md:grid @md:grid-cols-2 @md:gap-4">
+      <DataItemLabel>{label}</DataItemLabel>
+      <DataItemValue>{value}</DataItemValue>
+    </div>
+  );
+}
+
+ScoreDataItem.propTypes = {
+  label: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
+  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]).isRequired,
```
Comment (forresttanaka) [inline components/score-view-data-area.js:33]: I haven’t seen many people aside from me use `oneOfType` — nice! It can catch some good problems.

File: components/search-page-header.js [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+    if (path.endsWith(`#!chip`)) {
+      setButtonStyle("chip");
+    } else if (path.endsWith(`#!accessibility`)) {
+      setButtonStyle("accessibility");
+    } else {
+      setButtonStyle("score");
+    }
+  }, [router]);
+  return buttonStyle;
+}
```
Comment (forresttanaka) [inline components/search-page-header.js]: Again, make sure you an justify this. And I don’t mean to justify it to me here, but just to have the ability to explain why you use a custom hook in only one place. Maybe you expect to need it in other places in the future, or it just helps you keep things clear in your head.

File: components/search-page-header.js [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+  queryString: PropTypes.string.isRequired,
+  buttonStyle: PropTypes.string.isRequired,
+};
+
+function AccessibilityDataViewLink({ queryString, buttonStyle }) {
+  const editPath = `/search?${queryString}#!accessibility`;
+  let className = "bg-button-secondary text-button-secondary border-brand ";
+  if (buttonStyle === "accessibility") {
+    className = "text-button-primary outline-none bg-button-primary";
+  }
```
Comment (forresttanaka) [inline components/search-page-header.js]: If you use a conditional expression to assign these classes to `className`, you avoid assigning two different values to the same variable. That will increase this code’s performance by 0.000000001%. But no, it’s just unusual to see this operation this way. Alternatively, `let` and `if…else` also avoids assigning two different values to the same variable.

File: components/search-page-header.js [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+  queryString: PropTypes.string.isRequired,
+  buttonStyle: PropTypes.string.isRequired,
+};
+
+function ScoreViewLink({ queryString, buttonStyle }) {
+  const editPath = `/search?${queryString}`;
+  let className = "bg-button-secondary text-button-secondary border-brand ";
+  if (buttonStyle === "score") {
+    className = "text-button-primary outline-none bg-button-primary";
+  }
```
Comment (forresttanaka) [inline components/search-page-header.js]: Since you have the same code in both components, this might provide a good place to put this code in a shared function that takes `queryString` and returns `className`. Especially with Tailwind CSS className strings, I try to avoid repeating them in different places because it’s so easy to change one and forget to change the other, and a shared function helps.

File: components/search-page-header.js [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+      >
+        Score
+      </ButtonLink>
+    </div>
+  );
+}
+ScoreViewLink.propTypes = {
+  queryString: PropTypes.string.isRequired,
+  buttonStyle: PropTypes.string.isRequired,
+};
```
Comment (forresttanaka) [inline components/search-page-header.js:99]: Reading all of this now, could both of these components get combined into one? They only differ in the button label and text, and editPath. I could see, for documentation purposes, preserving these two components but making them stubs that call a shared component that parameterizes their differences, which could come down to just the label, text, and editPath if you still extract the top five lines into a separate function. It‘s arguable because these aren’t that big, in fact not as big as what I type here, but it seems a shame to have two components that do so much similarly.

File: components/snps-diagram.js [hunk 17]
```diff
@@ -1,7 +1,7 @@
 import PropTypes from "prop-types";
 import { DataAreaTitle, DataPanel } from "./data-area";
 
-const MIN_DISTANCE_BETWEEN_LABELS = 90;
+const MIN_DISTANCE_BETWEEN_LABELS = 94;
```
Comment (forresttanaka) [inline components/snps-diagram.js:4]: Yay! Converting these to constants has already paid for itself.

File: pages/query.js [hunk 18]
```diff
@@ -123,12 +128,30 @@ export default function Query() {
...
+                </Button>
+                <span> or </span>
+                <Button
+                  label="coordinates ranges"
+                  type="secondary"
+                  onClick={() => setTextInput(exampleCoordinates)}
+                >
+                  coordinates ranges
+                </Button>
+              </div>
```
Comment (forresttanaka) [inline pages/query.js]: You can also make the `<span>` elements `<div>` elements because they all exist within a `flex` div. That could help with vertical alignment issues, though if you don’t have any of those issues, it doesn‘t really matter.

File: pages/search.js [hunk 19]
```diff
@@ -20,10 +29,31 @@ import getSnpsInfo from "../lib/get-snps-info";
...
+  const router = useRouter();
+
+  useEffect(() => {
+    const path = router.asPath;
+    if (path.endsWith(`#!chip`) || path.endsWith(`#!accessibility`)) {
+      setShowScoreView(false);
+    }
+  }, [router]);
+  return showScoreView;
+}
```
Comment (forresttanaka) [inline pages/search.js:45]: It doesn’t look like you use this custom hook anywhere? It doesn’t generate an ESLint error for me for some reason, but I literally don’t see the string `useScore` anywhere else.

File: pages/search.js [hunk 20]
```diff
@@ -20,10 +29,31 @@ import getSnpsInfo from "../lib/get-snps-info";
...
 
-export default function Search({ data }) {
+export default function Search({ data, queryString }) {
   const [showMoreFreqs, setShowMoreFreqs] = useState(false);
+  const [showSoreView, setShowScoreView] = useState(true);
```
Comment (forresttanaka) [inline pages/search.js]: `showScoreView`

**Review by forresttanaka — CHANGES_REQUESTED** [review]
Looks like some good functionality here. Most of my comments are suggestions only, or are arguable, but found a couple errors.

With most of my days filled with code reviews, it really helps me take less time when the code is documented so I at least know what the purpose of functions or sections of functions are, and I suggest it helps you a few months from now after you forget what your code does. At least, this happens to me. We have several cases of code in encoded that we don’t know why it’s there, because the person who wrote it didn’t remember and didn’t comment it.

**Comment by mingjiecn** [comment]
> Looks like some good functionality here. Most of my comments are suggestions only, or are arguable, but found a couple errors.
> 
> With most of my days filled with code reviews, it really helps me take less time when the code is documented so I at least know what the purpose of functions or sections of functions are, and I suggest it helps you a few months from now after you forget what your code does. At least, this happens to me. We have several cases of code in encoded that we don’t know why it’s there, because the person who wrote it didn’t remember and didn’t comment it.

Thanks, Forrest! I will try to doc as much as possible.

File: components/accessibility-view.js (hunk 2 above)
Comment (mingjiecn) [reply components/accessibility-view.js]: Didn't realize that I don't need a custom hook. Thanks for the explanation!

File: components/chip-data-table.js (hunk 8 above)
Comment (mingjiecn) [reply components/chip-data-table.js:45]: Yes those are encode links, not internal. Thanks for checking!

File: components/chip-data-view.js (hunk 9 above)
Comment (mingjiecn) [reply components/chip-data-view.js:12]: added

File: components/search-page-header.js (hunk 16 above)
Comment (mingjiecn) [reply components/search-page-header.js:99]: You are absolutely right.

File: components/chip-data-view.js [hunk 21]
```diff
@@ -1,59 +1,51 @@
...
 
+// To dynamically load component ChipDataBar on the client side,
+// use the ssr option to disable server-rendering since ChipDataBar relies on browser APIs like window.
 const ChipDataBar = dynamic(() => import("./chip-data-bar-chart"), {
   ssr: false,
 });
 
-export function useChipData() {
+/**
+ * This is the view for diasplay ChIP-seq data for a variant.
```
Comment (forresttanaka) [inline components/chip-data-view.js]: “display”

**Review by forresttanaka — APPROVED** [review]
💙 Awesome! The new documentation helps a _ton_. I just have one comment misspelling that I have no problem if you let it go.

File: components/chip-data-view.js (hunk 21 above)
Comment (mingjiecn) [reply components/chip-data-view.js]: will fix it first
